presto-main/src/main/java/com/facebook/presto/operator/VariableWidthBlockEncodingBuffers.java

```diff
@@ -0,0 +1,257 @@
+                length += POSITION_SIZE;
+            }
+
+            rowSizes[i] += length;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendOffsetsAndSlices();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        output.writeInt(bufferedPositionCount);
+
+        // offsets
+        // note that VariableWidthBlock doesn't write the initial offset 0
+        if (offsetsBufferIndex > 0) {
+            output.appendBytes(offsetsBuffer, 0, offsetsBufferIndex);
+        }
+
+        // nulls
+        serializeNullsTo(output);
+
+        // slice
+        output.writeInt(sliceBufferIndex);  // totalLength
+        if (sliceBufferIndex > 0) {
+            output.appendBytes(sliceBuffer, 0, sliceBufferIndex);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPositionsRetainedSizeInBytes() +
+                sizeOf(offsetsBuffer) +
+                getNullsBufferRetainedSizeInBytes() +
+                sizeOf(sliceBuffer);
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                offsetsBufferIndex +            // offsets buffer.
+                SIZE_OF_INT +                   // sliceBuffer size.
+                sliceBufferIndex +               // sliceBuffer
+                getNullsBufferSerializedSizeInBytes();  // nulls
+    }
+
+    // This implementation requires variableWidthBlock.getPositionOffset() to be public but it's much faster than the below one
+    private void appendOffsetsAndSlices()
+    {
+        ensureOffsetsBufferCapacity();
+
+        AbstractVariableWidthBlock variableWidthBlock = (AbstractVariableWidthBlock) decodedBlock;
+        int[] positions = getPositions();
+
+        // Get the whole slice then access specific rows to allow faster copy
+        int sliceLength = variableWidthBlock.getPositionOffset(variableWidthBlock.getPositionCount()) - variableWidthBlock.getPositionOffset(0);
+
+        byte[] sliceBase = (byte[]) variableWidthBlock.getSlice(0, 0, sliceLength).getBase();
+
+        for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+            int position = positions[i];
+            int beginOffset = variableWidthBlock.getPositionOffset(position);
+            int endOffset = variableWidthBlock.getPositionOffset(position + 1);
+            int currentRowSize = endOffset - beginOffset;
+
+            lastOffset += currentRowSize;
+
+            offsetsBufferIndex = setIntUnchecked(offsetsBuffer, offsetsBufferIndex, lastOffset);
+
+            ensureSliceBufferCapacity(currentRowSize);
```

Comment:
- move this inside the `if` block
- this call is confusing because it passes additional capacity request as opposed to total capacity like other similar methods; let's consolidate all these methods in a single place and make them consistent

Reply:
Created separate PR https://github.com/prestodb/presto/pull/13284 Move Arrays class to presto-array
And consolidated all ensureCapacity calls to use Arrays class.